README.md
# Mapcrafter [![Build Status](https://travis-ci.org/miclav/mapcrafter.svg?branch=master)](https://travis-ci.org/miclav/mapcrafter)

by Moritz Hilscher *(modified by Michael Lavaire)*

[![dockeri.co](https://dockeri.co/image/miclav/mapcrafter)](https://hub.docker.com/r/miclav/mapcrafter)

Mapcrafter is a high performance Minecraft map renderer written in C++. It renders
Minecraft worlds to a bunch of tiles (images) which are viewable as a map in any browser
using Leaflet.js.


<!-- TOC -->

- [Block status](#block-status)
- [Features](#features)
- [Requirements](#requirements)
- [Example maps](#example-maps)
- [Help](#help)
  - [Documentation](#documentation)
  - [Reporting bugs and feature requests](#reporting-bugs-and-feature-requests)
  - [Community help](#community-help)
- [Acknowledgements](#acknowledgements)
- [License](#license)
- [Version history](#version-history)

<!-- /TOC -->

## Block Status
Featuring blocks up to 1.20
![](./src/data/blocks/isometric_0_16.png)

![](./docs/mapcrafter.png)

## Supported Minecraft versions

The current version supports Minecraft versions 1.18.x. Newer versions may partially work.
For Minecraft 1.17.x (or older) please use previous version of Mapcrafter.

## Features

- **Web output:** Render your Minecraft worlds to maps viewable in any web browser!
- **Different render views:** Choose between different perspectives to render your world
  from! A 2D topdown and a 3D isometric render view are available!
- **Different rotations:** Choose from four different rotations to render your worlds from!
- **Different render modes:** Choose between different render modes like day, night and cave
  for your maps!
- **Different overlays:** Show additional information on your map! For example: Where can
  slimes spawn? Where can monsters spawn at night?
- **Configuration files:** Highly-customizable which worlds are rendered with which render
  view and other render parameters!
- **Markers:** Automatically generated markers from your Minecraft world data!
- **Other stuff:** Biome colors, incremental rendering, multithreading

## Requirements

- A Linux-based or Unix-like operating system like Mac OS
- A decent C++ compiler (preferable `gcc` >= 4.4, or `clang`), CMake and make if you want
  to build Mapcrafter from source
- Some libraries:
  - `libpng`
  - `libjpeg` (but you may use libjpeg-turbo as drop in replacement)
  - `libboost-iostreams`
  - `libboost-system`
  - `libboost-filesystem` (>= 1.42)
  - `libboost-program-options`
  - `libboost-test` (if you want to use the tests)
- For your Minecraft worlds:
  - Anvil world format
  - Minecraft resource packs

## Example maps

There are a few example maps of the renderer on the
[GitHub Wiki](https://github.com/mapcrafter/mapcrafter/wiki/Example-maps).
Please feel free to add your own maps to this list.

## Help

### Documentation

You can read the current documentation on [docs.mapcrafter.org](http://docs.mapcrafter.org)
and you can also [download other builds](https://readthedocs.org/projects/mapcrafter/downloads/)
from there. The documentation source is in the `docs/` directory and you can build it
yourself with Sphinx.

### Reporting bugs and feature requests

If you find bugs or problems when using Mapcrafter or if you have ideas for new
features, then please feel free to add an issue to the [GitHub issue
tracker](https://github.com/mapcrafter/mapcrafter/issues).

### Community help

[Join us on Discord](https://discord.gg/QNe8jXT), where you can meet other Mapcrafter users
and ask for help. Please be respectful with everyone and have patience with people who are
offering their time to help!

## Acknowledgements

Thanks to pigmap and Minecraft Overviewer, whose documentations and source code
were very helpful. I also used the alpha blending code of pigmap and some maps
stuff of the template from Minecraft Overviewer.

## License

Mapcrafter is free software and available under the GPL license.  You can
access the latest source code of Mapcrafter on GitHub:
http://github.com/mapcrafter/mapcrafter

## Version history

See the [CHANGELOG](./CHANGELOG.md) for details.


CHANGELOG.md
# Mapcrafter version history

**v.3.1.388**

* Blocks updated (1.20.3-pre2 & 23w46a)
* Fix UI coordinates (Thanks to @kevinwchang)
* Fix signs from 1.20 in marker tool (Thanks to @doismellburning)

**v.3.1.328**

* 1.19-pre1 blocks included
* Blocks updated with a lot of texture mapping issues resolved in a Blockcrafter update

**v.3.1.323**

* Support for 1.18.x
  * Dockerfile build fixed
  * Entity cache file not saved in world folder anymore. World folder can now be mounted as read only.
  * Breaking change, doesn't support the old chunk format anymore (sorry, :( no time to develop a dynamic convertion of old Minecraft file format)
  * Chunks height supoprted from -64 to 319 included.
  * Support new biome palette
  * Biomes now identified by their name, no more by ID
* Add support for alternate block rotation, improving overall visual, breaking monotonous flat surfaces
* Block atlas, now palettized to minimize memory footprint and disk space
* Performance improvements:
  * Region header IO operations reduced from 4098 to 4 calls.
  * Rendering reworked by separating block render and water layer
  * World rotation now supported on the rendering layer, no more world data rotation
* Bug fixes:
  * Side view of waterloged block wasn't rendering side water
* Still to be worked on:
  * lighting on edges funky
  * few water rendering edge cases
  * Blockcrafter:
    * Textures misplaced on some blocks (ex: cauldron, glass, ...)
    * face sorting in blockcrafter to minimize Z-fighting
    * Missing block types:
      * Banners (doable?)
      * end gate
      * standard heads (skeleton, player, wither skeleton, creeper, dragon)
      * Candle + cake, faces not properly placed
      * Lit candles don't have flames
    * UV's rotation not right in some circonstances

**v.3.0**

* Add support for Minecraft versions 1.13 through 1.16 (#279)

**v.2.4**

* Add support for Minecraft 1.11 and 1.12 blocks (thanks to Bjarno #256)

**v.2.3.1**

* Fixed wrong Leaflet coordinate conversion in topdown and isometric render view when
    using different tiles sizes and fixed wrong chunk <-> tile mapping which caused the
    issue in isometric render view (#198)
  * **Unfortunately you have to force-re-render your isometric maps as the isometric
    chunk <-> tile mapping has slightly changed.**

**v.2.3**

* Added support for Minecraft 1.10 blocks (#226)
* Added new bootstrap web UI (thanks to jamestaylr #205)
* Added `postfix` option for signs to also use a postfix instead of just a prefix (#223)
* Added `--force-render-all` to force-render all maps (#212)
* Fixed map selection on mobile devices (thanks to joshproehl #220 #123)
* Fixed a bug in the bilinear image resizing algorithm that caused dark sides of
    grass and other transparent blocks

**v.2.2.1**

* Updated Minecraft version file to 1.9

**v.2.2**

* Added support for 1.9 blocks (and hoppers)
* Fixed bug which caused water not being rendered (#178)
* Fixed circular crop overflowing bug (#179)
* Fixed a config validation bug (#185)

**v.2.1**

* Added support for corner stairs (thanks for some Minecraft Overviewer code!)
* Added signs to isometric and topdown render view
* Fixed a lighting bug that appeared using some Debian packages
* Fixed wrong slime overlay
* Fixed rotation of pumpkins in isometric render view

**v.2.0**

* Huge internal code refactoring™
* Support for multiple rendering engines: A 2D topdown and the usual 3D isometric render
  view are available
* Support for indexed png images, reducing size of output images drastically
* New overlay render modes (slime chunks, show where monsters can spawn at day/night)
* Configurable tile sizes (to improve performance for small texture sizes)
* Configurable water opacity/transparency
* Restructured internal render management components to allow integration of Mapcrafter
  rendering core into different applications
* Further integration of Mapcrafter into a graphical user interface
  ([mapcrafter-gui](https://github.com/mapcrafter/mapcrafter-gui))
* **There is a new repository for Debian/Ubuntu packages. Please have a look at the
    documentation and update your apt source listings!**

**v.1.5.4**

* Added a detection in the CMake files which compiler C++11 flag to use (for newer
  compilers which do not support the older ``-std=c++0x`` flag)
* Fixed bug with the marker tool (new json sign text format was not parsed properly)
* Fixed bug in texture extraction script (convert command was not properly detected)
* Fixed bug with some textures (#122) by making Mapcrafter able to read indexed PNG images
  and PNG images with different other color formats

**v.1.5.3**

* Better handling of corrupt world files (for now, bigger refactoring of that probably later)
* Added a fix using imagemagick in texture extraction script to prevent libpng warnings
* Fixed bug with boost program options which made it unable to use paths
  containing spaces for the ``--config`` command line option
* Fixed wrong error messages when loading configuration files (Mapcrafter tried
  to read not existing files and even directories if specified as configuration file)

**v.1.5.2**

* Fixed the marker tool to work with the new internal Minecraft 1.8 sign format
* Fixed bug with global logging configuration file installed and searched in wrong directory
* Fixed bug causing the Leaflet map being stuck
* Fixed some small issues with new 1.8 blocks

**v.1.5.1**

* Added ability to configure logging facility with configuration files
* Added colored terminal output for warning/error log messages
* Improved cave rendermode with a new high contrast block coloring (thanks to dtfinch, #100),
  old behavior is still available with a configuration option for compatibility reasons
* Added support for Minecraft 1.8
  * **Since the structure of the textures has changed a bit (especially chest textures),
    you should extract the texture files from a 1.8 Minecraft Jar file to a new, clean
    texture directory.**

**v.1.5**


src/mapcraftercore/mc/java.cpp
/*
 * Copyright 2012-2016 Moritz Hilscher
 *
 * This file is part of Mapcrafter.
 *
 * Mapcrafter is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Mapcrafter is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Mapcrafter.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "java.h"

namespace mapcrafter {
namespace mc {

JavaRandom::JavaRandom()
	: seed(0) {
}

JavaRandom::~JavaRandom() {
}

void JavaRandom::setSeed(long long seed) {
	this->seed = (seed ^ 0x5DEECE66DLL) & ((1LL << 48) - 1);
}

int JavaRandom::next(int bits) {
	seed = (seed * 0x5DEECE66DLL + 0xBLL) & ((1LL << 48) - 1);
	return (int) (seed >> (48 - bits));
}

int JavaRandom::nextInt(int max) {
	// invalid max
	if (max <= 0)
		return -1;
	if ((max & -max) == max) // i.e., n is a power of 2
		return (int) ((max * (long long) next(31)) >> 31);
	int bits, val;
	do {
		bits = next(31);
		val = bits % max;
	} while (bits - val + (max - 1) < 0);
	return val;
}

const int JavaSimplexGenerator::GRADIENTS[12][3] = {
	{ 1, 1, 0},
	{-1, 1, 0},
	{ 1,-1, 0},
	{-1,-1, 0},
	{ 1, 0, 1},
	{-1, 0, 1},
	{ 1, 0,-1},
	{-1, 0,-1},
	{ 0, 1, 1},
	{ 0,-1, 1},
	{ 0, 1,-1},
	{ 0,-1,-1}
};

// If you are an Mojang/Microsoft employee reading this:
// These numbers appeared to me in a dream
// I don't know how, but they help to reproduce the swamp grass colors just like in Minecraft ;)
const int JavaSimplexGenerator::PERMUTATIONS[512] = {
	64, 175, 124, 148, 10, 239, 244, 91, 138, 73, 228, 171, 27, 134, 77, 122,
	238, 196, 202, 181, 211, 7, 49, 173, 48, 165, 120, 217, 129, 56, 153, 8,
	140, 141, 21, 130, 71, 100, 132, 23, 176, 250, 29, 104, 149, 159, 180, 237,
	247, 11, 252, 241, 14, 2, 219, 75, 178, 151, 233, 251, 103, 45, 52, 201,
	222, 18, 223, 88, 136, 34, 227, 235, 35, 160, 0, 131, 51, 214, 39, 216,
	207, 26, 137, 185, 41, 13, 249, 54, 112, 5, 66, 242, 157, 158, 28, 89, 86,
	192, 172, 17, 69, 204, 38, 221, 65, 166, 9, 226, 33, 30, 84, 240, 59, 224,
	127, 108, 92, 146, 99, 195, 255, 98, 126, 4, 133, 236, 189, 121, 144, 183,
	80, 109, 191, 218, 161, 53, 25, 93, 72, 150, 163, 234, 205, 152, 61, 37,
	197, 78, 81, 32, 85, 70, 187, 63, 96, 115, 117, 184, 139, 79, 74, 46, 188,
	182, 76, 31, 174, 57, 68, 198, 90, 245, 230, 106, 94, 212, 190, 16, 200,
	213, 206, 44, 43, 215, 231, 12, 177, 203, 220, 24, 170, 19, 209, 82, 95,
	125, 194, 248, 208, 55, 67, 1, 87, 110, 135, 162, 128, 3, 60, 225, 15, 186,
	232, 145, 119, 142, 113, 154, 102, 164, 42, 156, 210, 22, 253, 147, 169,
	193, 83, 143, 118, 123, 254, 167, 111, 114, 6, 50, 40, 199, 179, 246, 20,
	107, 168, 97, 229, 101, 155, 62, 47, 58, 116, 243, 105, 36, 64, 175, 124,
	148, 10, 239, 244, 91, 138, 73, 228, 171, 27, 134, 77, 122, 238, 196, 202,
	181, 211, 7, 49, 173, 48, 165, 120, 217, 129, 56, 153, 8, 140, 141, 21,
	130, 71, 100, 132, 23, 176, 250, 29, 104, 149, 159, 180, 237, 247, 11, 252,
	241, 14, 2, 219, 75, 178, 151, 233, 251, 103, 45, 52, 201, 222, 18, 223,
	88, 136, 34, 227, 235, 35, 160, 0, 131, 51, 214, 39, 216, 207, 26, 137,
	185, 41, 13, 249, 54, 112, 5, 66, 242, 157, 158, 28, 89, 86, 192, 172, 17,
	69, 204, 38, 221, 65, 166, 9, 226, 33, 30, 84, 240, 59, 224, 127, 108, 92,
	146, 99, 195, 255, 98, 126, 4, 133, 236, 189, 121, 144, 183, 80, 109, 191,
	218, 161, 53, 25, 93, 72, 150, 163, 234, 205, 152, 61, 37, 197, 78, 81, 32,
	85, 70, 187, 63, 96, 115, 117, 184, 139, 79, 74, 46, 188, 182, 76, 31, 174,
	57, 68, 198, 90, 245, 230, 106, 94, 212, 190, 16, 200, 213, 206, 44, 43,
	215, 231, 12, 177, 203, 220, 24, 170, 19, 209, 82, 95, 125, 194, 248, 208,
	55, 67, 1, 87, 110, 135, 162, 128, 3, 60, 225, 15, 186, 232, 145, 119, 142,
	113, 154, 102, 164, 42, 156, 210, 22, 253, 147, 169, 193, 83, 143, 118,
	123, 254, 167, 111, 114, 6, 50, 40, 199, 179, 246, 20, 107, 168, 97, 229,
	101, 155, 62, 47, 58, 116, 243, 105, 36
};


namespace {

double dot(const int* v, double x, double y) {
	return (double) v[0] * x + (double) v[1] * y;
}

}

double JavaSimplexGenerator::getValue(double x, double y) const {
	// Once again, have a look at:
	// http://staffwww.itn.liu.se/~stegu/simplexnoise/SimplexNoise.java
	
	double s = (x + y) * F2;
	int i = std::floor(x + s);
	int j = std::floor(y + s);

	double t = (double) (i + j) * G2;
	double X0 = (double)i - t;
	double Y0 = (double)j - t;
	double x0 = x - X0;
	double y0 = y - Y0;
	
	int i1, j1;
	if (x0 > y0) {
		i1 = 1;
		j1 = 0;
	} else {
		i1 = 0;
		j1 = 1;
	}

	double x1 = x0 - (double) i1 + G2;
	double y1 = y0 - (double) j1 + G2;
	double x2 = x0 - 1.0 + 2.0 * G2;
	double y2 = y0 - 1.0 + 2.0 * G2;
	int ii = i & 255;
	int jj = j & 255;
	int gi0 = PERMUTATIONS[ii + PERMUTATIONS[jj]] % 12;
	int gi1 = PERMUTATIONS[ii + i1 + PERMUTATIONS[jj + j1]] % 12;
	int gi2 = PERMUTATIONS[ii + 1 + PERMUTATIONS[jj + 1]] % 12;
	
	double t0 = 0.5 - x0 * x0 - y0 * y0;
	double n0;
	if (t0 < 0.0) {
		n0 = 0.0;
	} else {
		t0 = t0 * t0;
		n0 = t0 * t0 * dot(GRADIENTS[gi0], x0, y0);
	}

	double t1 = 0.5 - x1 * x1 - y1 * y1;
	double n1;
	if (t1 < 0.0) {
		n1 = 0.0;
	} else {
		t1 = t1 * t1;
		n1 = t1 * t1 * dot(GRADIENTS[gi1], x1, y1);
	}

	double t2 = 0.5 - x2 * x2 - y2 * y2;
	double n2;
	if (t2 < 0.0) {
		n2 = 0.0;
	} else {
		t2 = t2 * t2;
		n2 = t2 * t2 * dot(GRADIENTS[gi2], x2, y2);
	}

	return 70.0 * (n0 + n1 + n2);
}

}
}

